README.md
# Interlude [![Build Status](https://secure.travis-ci.org/clux/interlude.png)](http://travis-ci.org/clux/interlude)
Interlude is ES5 JavaScript with Haskell inspired enhancements.
It's aims to simplify and abstract common patterns by joining
common higher order functions with the ES5 arsenal to allow a
more declarative style with negligible efficiency changes.

It does not simply alias ES5 methods, and it does not touch prototypes.
*It curries*.

## Usage
Attach it to the short variable of choice:

```javascript
var $ = require('interlude');
```

Then spice up your JavaScript with some functional-style programming;

```javascript
[1,3,2,6,5,4].filter($.gt(4));
// [ 6, 5 ]

$.range(5).map($.pow(2));
// [ 1, 4, 9, 16, 25 ]

var nested = [ [1,3,2], [2,2], [1,4,2,3] ];
nested.filter($.all($.eq(2)));
// [ [2, 2] ]

// outer sort by property
nested.sort($.comparing('length'));
// [ [ 2, 2 ], [ 1, 3, 2 ], [ 1, 4, 2, 3 ] ]

// inner sort by value
nested.map($.invoke('sort', $.compare(-1)));
// [ [ 2, 2 ], [ 3, 2, 1 ], [ 4, 3, 2, 1 ] ]

// used with arrays of pairs
var pairs = [['woo', 3], ['wee', 1], ['boo', 2]];
pairs.sort($.comparing(1)).map($.first); // [ 'wee', 'boo', 'woo' ]

$.zipWith($.plus3, [1,1,1,1,1], $.range(5), [1,0,0]);
// [ 3, 3, 4 ]

// Powers of two
$.iterate(8, 2, $.times(2));
// [ 2, 4, 8, 16, 32, 64, 128, 256 ]


// Pascal's Triangle
var pascalNext = function (row) {
  return $.zipWith($.plus2, row.concat(0), [0].concat(row));
};
$.iterate(6, [1], pascalNext);
// [ [ 1 ],
//   [ 1, 1 ],
//   [ 1, 2, 1 ],
//   [ 1, 3, 3, 1 ],
//   [ 1, 4, 6, 4, 1 ],
//   [ 1, 5, 10, 10, 5, 1 ] ]

// Prime numbers
var notCoprime = $($.gcd, $.gt(1));
$.nubBy(notCoprime, $.range(2, 20));
// [ 2, 3, 5, 7, 11, 13, 17, 19 ]
```

Interlude is merely a stable front for three re-exported modules:

```
interlude
├─── operators
├─── autonomy
└─── subset
```

These modules are of course requirable by themselves, and we encourage you to require them directly. The submodules are small (<300 lines each) so by requiring one of them, you are less likely to have to search around to wonder why you included 1000-some lines when you just wanted 2-3 functions.

Minimizing dependencies make it easier for people to reason about your code. That said, this module is here when/if you want it all under the same namespace.

Additionally, interlude provides a stable front for the three submodules which are still pre 1.0.

Regardless, you should read their short and independent APIs:

- [operators](https://github.com/clux/operators) ([API](https://github.com/clux/operators/blob/master/api.md))
- [autonomy](https://github.com/clux/autonomy) ([API](https://github.com/clux/autonomy/blob/master/api.md))
- [subset](https://github.com/clux/subset) ([API](https://github.com/clux/subset/blob/master/api.md))

Additionally, two extra modules (which were not included in interlude due to their smaller likelihood of use) are perfect fits, does not overlap in API, and are highly recommended *when needed*. Read their short APIs too:

- [wrappers](https://github.com/clux/wrappers) ([API](https://github.com/clux/wrappers/blob/master/api.md))
- [typr](https://github.com/clux/typr) ([API](https://github.com/clux/typr/blob/master/api.md))

## Installation

```bash
$ npm install interlude
```

## Running tests
Install development dependencies

```bash
$ npm install
```

Run the tests

```bash
$ npm test
```

## License
MIT-Licensed. See LICENSE file for details.


History.md
1.0.2 / 2012-11-12
==================
  * extend a blank copy of `seq` rather than require result of `autonomy`.
  (Was overwriting the cached require results of autonomy in linked installs)

1.0.1 / 2012-11-11
==================
  * `$()` now a shortcut for deprecated `$.seq`

1.0.0 / 2012-10-24
==================
  * retroactive inclusion of an overloaded version `$.compare` from subset

1.0.0 / 2012-10-20
==================
  * interlude is now a stable front for autonomy, subset and operators
  * `has`, `seq2`, `seq3`, `seq4`, `getDeep` removed from autonomy
  * `replicate` is now array safe
  * `get` is now variadic and assumes the role of `getDeep`
  * `wrappers` excluded from exports because:
    - less stable API
    - less likelihood of use
  * `subset` includes `isSubsetOf` (retroactively available to previous versions)

0.7.0 / 2012-07-04
==================
  * autonomy updated to 0.3.0 changing `scan` argument order
  * allow use on any node version

0.6.0 / 2012-07-03
==================
  * wrappers updated to 0.2.0 removing `either` and `guard`

0.5.1 / 2012-06-30
==================
  * include new (v0.2.0) autonomy which comes without operators + pow/log but with extend
  * operators included directly (now comes with pow/log + pow2/log2)
  * improve DOCS browsability and just api in general

0.5.0 / 2012-06-22
==================
  * update wrappers to 0.1.0 (now close to done, has a lot more)

0.4.0 / 2012-06-17
==================
  * use latest operators, 0.3 missed some extras
  * comparing/compare direction parameter is now the factor +1/-1 with +1 omittable
  * set operations + comparison/equality generalizations moved to `subset` module
  * everything else (must haves) moved into `origin` module, which re-exports operators.
  * trace/traceBy/wrap/once/memoize initial wrappers included via `wrappers` module
  * module basically just a re-export


0.3.0 / 2012-06-15
==================
  * `operators` moved to own module (which now is bigger and better)
  * renamed `collect` to `pluck` to conserve (at least some) conventions
  * $.equality, $.comparing $.compare added
  * removed awful setters
  * delete/insert/deleteBy/insertBy now all modify input (insertBy didn't)
  * delete/deleteBy now only deletes first occurrence
  * difference/differenceBy introduced (non-modifying)
  * Data.List like API now dependency free, can be factored out
  * compose redone for performance and sensibility. See: `seq`, `seq2`, `seq3`, `seq4`.
  * experimental stuff removed from module.exports while it's in the works
  * tests moved from expresso to tap

0.2.0 / 2012-06-03
==================
  * Data.List generic insert/delete/group/union/maximum/minimum/nub done

0.1.0 / 2012-06-02
==================
  * First proper version under 'interlude'


test/utils.js
var tap = require('tap')
  , test = tap.test
  , $ = require('../');

test("readme", function (t) {
  t.deepEqual([1,3,2,6,5,4].filter($.gt(4)), [ 6, 5 ], "filter gt 4");

  t.deepEqual($.range(5).map($.pow(2)), [ 1, 4, 9, 16, 25 ], "[1..5] map pow 2");

  var nested = [ [1,3,2], [2,2], [1,4,2,3] ];
  t.deepEqual(nested.filter($.all($.eq(2))), [ [2, 2] ], "nested filter all eq 2");

  nested.sort($.comparing('length'));
  t.deepEqual(nested, [ [ 2, 2 ], [ 1, 3, 2 ], [ 1, 4, 2, 3 ] ], "nested sort prop");

  nested.map($.invoke('sort', $.compare(-1)));
  t.deepEqual(nested, [ [ 2, 2 ], [ 3, 2, 1 ], [ 4, 3, 2, 1 ] ], "nested sort value");

  var pairs = [['woo', 3], ['wee', 1], ['boo', 2]];
  var pairSort = pairs.sort($.comparing(1)).map($.first);
  t.deepEqual(pairSort, [ 'wee', 'boo', 'woo' ], "pairSort map first");

  var zipP3 = $.zipWith($.plus3, [1,1,1,1,1], $.range(5), [1,0,0]);
  t.deepEqual(zipP3, [ 3, 3, 4 ], "zipWith +_3");

  t.deepEqual($.iterate(8, 2, $.times(2)), [ 2, 4, 8, 16, 32, 64, 128, 256 ], "Powers of two");

  // Pascal's Triangle
  var pascalNext = function (row) {
    return $.zipWith($.plus2, row.concat(0), [0].concat(row));
  };
  var pascal = $.iterate(6, [1], pascalNext);
  t.deepEqual(pascal[0], [ 1 ], "pascal[0]");
  t.deepEqual(pascal[1], [ 1, 1 ], "pascal[1]");
  t.deepEqual(pascal[2], [ 1, 2, 1 ], "pascal[2]");
  t.deepEqual(pascal[3], [ 1, 3, 3, 1 ], "pascal[3]");
  t.deepEqual(pascal[4], [ 1, 4, 6, 4, 1 ], "pascal[4]");
  t.deepEqual(pascal[5], [ 1, 5, 10, 10, 5, 1 ], "pascal[5]");
  t.equal(pascal.length, 6, "got exactly the number of rows we asked for in pascal");

  // Prime numbers
  var notCoprime = $($.gcd, $.gt(1));
  var primes = $.nubBy(notCoprime, $.range(2, 20));
  t.deepEqual(primes, [ 2, 3, 5, 7, 11, 13, 17, 19 ], "primes <= 20");

  t.end();
});


operators.js
var $ = {}
  , reduce = Array.prototype.reduce
  , concat = Array.prototype.concat;

// multi-parameter versions for the associative operators:
$.plus2 = function (x, y) {
  return x + y;
};
$.plus3 = function (x, y, z) {
  return x + y + z;
};
$.plus4 = function (x, y, z, w) {
  return x + y + z + w;
};

$.times2 = function (x, y) {
  return x * y;
};
$.times3 = function (x, y, z) {
  return x * y * z;
};
$.times4 = function (x, y, z, w) {
  return x * y * z * w;
};

$.and2 = function (x, y) {
  return x && y;
};
$.and3 = function (x, y, z) {
  return x && y && z;
};
$.and4 = function (x, y, z, w) {
  return x && y && z && w;
};

$.or2 = function (x, y) {
  return x || y;
};
$.or3 = function (x, y, z) {
  return x || y || z;
};
$.or4 = function (x, y, z, w) {
  return x || y || z || w;
};

$.append2 = function (xs, ys) {
  return xs.concat(ys);
};
$.append3 = function (xs, ys, zs) {
  return xs.concat(ys, zs);
};
$.append4 = function (xs, ys, zs, ws) {
  return xs.concat(ys, zs, ws);
};

// makes sense to have 3,4 parameter versions of this despite it being associative
$.prepend2 = function (xs, ys) {
  return ys.concat(xs);
};


// Array versions of 3/6 associative operators's reductions
$.sum = function (xs) {
  return xs.reduce($.plus2, 0);
};
$.product = function (xs) {
  return xs.reduce($.times2, 1);
};
$.flatten = function (xs) {
  return concat.apply([], xs);
};

// Cheapest general associative reductions without looping over arguments manually
// http://jsperf.com/reduce-vs-slicereduce3 <- the cost of abstractions
// any/all names more useful for accessors for Array::some/every
$.add = function () {
  return reduce.call(arguments, $.plus2, 0);
};
$.multiply = function () {
  return reduce.call(arguments, $.times2, 1);
};
$.concat = function () {
  return concat.apply([], arguments);
};


// non-associative operators only get the 2 argument version
$.minus2 = function (x, y) {
  return x - y;
};

$.divide2 = function (x, y) {
  return x / y;
};

$.div2 = function (x, y) {
  return Math.floor(x / y);
};

$.mod2 = function (x, y) {
  return x % y;
};

$.eq2 = function (x, y) {
  return x === y;
};

$.neq2 = function (x, y) {
  return x !== y;
};

$.gt2 = function (x, y) {
  return x > y;
};

$.lt2 = function (x, y) {
  return x < y;
};

$.gte2 = function (x, y) {
  return x >= y;
};

$.lte2 = function (x, y) {
  return x <= y;
};

// curried versions
$.plus = function (y) {
  return function (x) {
    return x + y;
  };
};

$.minus = function (y) {
  return function (x) {
    return x - y;
  };
};

$.times = function (y) {
  return function (x) {
    return x * y;
  };
};

$.divide = function (y) {
  return function (x) {
    return x / y;
  };
};

$.div = function (y) {
  return function (x) {
    return Math.floor(x / y);
  };
};

$.mod = function (y) {
  return function (x) {
    return x % y;
  };
};

$.append = function (ys) {
  return function (xs) {
    return xs.concat(ys);
  };
};

$.prepend = function (ys) {
  return function (xs) {
    return ys.concat(xs);
  };
};

$.gt = function (y) {
  return function (x) {
    return x > y;
  };
};

$.lt = function (y) {
  return function (x) {
    return x < y;
  };
};

$.eq = function (y) {
  return function (x) {
    return x === y;
  };
};

$.neq = function (y) {
  return function (x) {
    return x !== y;
  };
};

$.gte = function (y) {
  return function (x) {
    return x >= y;
  };
};

$.lte = function (y) {
  return function (x) {
    return x <= y;
  };
};

module.exports = $;


functional.js
var slice = Array.prototype.slice;
var f = {};

// common
var id = f.id = function (x) {
  return x;
};

var noop = f.noop = function () {
};

var constant = f.constant = function (val) {
  return function () {
    return val;
  };
};

// curried helper functions for common operations
f.plus = function (x) {
  return function (y) {
    return y + x;
  };
};

f.times = function (x) {
  return function (y) {
    return y * x;
  };
};

f.append = function (xs) {
  return function (ys) {
    return ys.concat(xs);
  };
};

f.compose = function (f) {
  return function (g) {
    return function (x) {
      return f(g(x));
    };
  };
};



// fold - Array::reduce with array curried
var fold = f.fold = function (f, initial) {
  return function (xs) {
    return xs.reduce(f, initial);
  };
};

// binary operators that will be folded over - not exported
var add = function (x, y) {
  return x + y;
};

var multiply = function (x, y) {
  return x * y;
};

var concat = function (xs, ys) {
  return xs.concat(ys);
};

var both = function (x, y) {
  return x && y;
};

var either = function (x, y) {
  return x || y;
};

var chain = function (f, g) {
  return function (x) {
    return f(g(x));
  };
};


// using fold with binary operators to give lifted functions
// any of these (i.e. they act on a list) can be remembered by "take the " fnName
// wheras the variadic counterpart has the name of the associative operator
f.sum = fold(add, 0);
f.product = fold(multiply, 1);
f.flat = fold(concat, []);
f.composition = fold(chain, id);

// we could export these
//f.and = fold(both, true);
//f.or = fold(either, false);

// but prefer faster ES5 methods
// best semantically; "we take the logical" and/or of list
f.and = function (f) {
  return function (xs) {
    return xs.every(f);
  };
};

f.or = function (f) {
  return function (xs) {
    return xs.some(f);
  };
};

// lifts and unlifts:
// because some functions are already variadic

// lift a multi param function f into a single array fn, inverse of unlift
var lift = f.lift = function (f, context) {
  return function (xs) {
    return f.apply(context, xs);
  };
};

// examples
// follows the "take the" fnName of list semantic
// max/min already have unlifted variadic counterparts so they can simply be lifted
var maximum = f.maximum = lift(Math.max, Math);
var minimum = f.maximum = lift(Math.min, Math);


// (new ClassInst) can't be lifted - has to be one manually
// maybe not want to put this in here, after all Object.create exists
// and classes are a different thing for people to disagree over
f.construct = function (Ctor, args) {
  var F = function () {
    Ctor.apply(this, args);
  };
  F.prototype = Ctor.prototype;
  return new F();
};


// take a function operating on an array and turn it into a multi-parameter function, inverse of lift
// because argument based functions are semantic tand is most sensible with zipWith
var unlift = f.unlift = function (f, context) {
  return function () {
    var args = slice.call(arguments, 0);
    return f.apply(context, [args]);
  };
};

// variadic version of lift functions
// named after the operation, i.e. "we " fnName (together) "arg1, arg2, ..."
f.add = unlift(f.sum);
f.multiply = unlift(f.product);
f.concat = unlift(f.flatten);
f.and = unlift(f.all);
f.or = unlift(f.any);
f.compose = unlift(f.composition);


// getters/setters

// simple proprety accessor
// can also be used for array number accessor
// get :: Int/String -> Property
f.get = function (prop) {
  return function (el) {
    return el[prop];
  };
};

// property get map -- equivalent to _.pluck or map(get('prop'))
// zip(pmap('length'), [[1,3,2],[2,1]], [[1],[2]], [[2,3],[2,1]]) -> [[3,1,2], [2,1,2]]
// pmap :: String -> [a] -> [b] -- both curried and as a two param fn
var pmap = f.pmap = function (propName, ary) {
  var fn = function (xs) {
    var result = [];
    for (var i = 0; i < xs.length; i += 1) {
      result[i] = xs[i][propName];
    }
    return result;
  };
  return (ary == null) ? fn : fn(ary);
};

f.set = function (propName) {
  return function (el, value) {
    el[propName] = value;
    return el;
  };
};

// modify a list of objects by setting propName on all objects to valFn(currObj)
// property set map -- equivalent to map(set('prop'))
var mmap = f.mmap = function (propName, valFn) {
  return function (xs) {
    for (var i = 0; i < xs.length; i += 1) {
      xs[i][propName] = valFn(xs[i]);
    }
    return xs;
  };
};

// can use mmap('prop1', constant(5))([{}, {a:2}]) -> [{prop1:5}, {a:2, prop1: 5}]|



// general map/zip helpers

// equivalent to _.range
f.range = function (start, stop, step) {
  if (arguments.length <= 1) {
    stop = start || 0;
    start = 0;
  }
  step = arguments[2] || 1;

  var len = Math.max(Math.ceil((stop - start) / step), 0)
    , idx = 0
    , range = new Array(len);

  while (idx < len) {
    range[idx++] = start;
    start += step;
  }

  return range;
};

f.iterate = function (f, times) {
  return function (x) {
    var result = [x];
    for (var i = 1; i < times; i += 1) {
      result.push(f(result[i - 1]));
    }
    return result;
  };
};


//TODO: throttle, memoize, debounce, once
//TODO: clone, extend

// can act as zipWith, zipWith3, zipWith4...
// zipper function must have the same number of arguments as there are lists
// but beyond that, it's very dynamic
// zipWith(function(x,y,z){return x+y+z;}, [1,3,2], [21,1], [2,3]) -> [24,7]
f.zipWith = function () {
  var fn = arguments[0]
    , args = slice.call(arguments, 1)
    , numLists = args.length
    , results = []
    , len = minimum(pmap('length', args));

  for (var i = 0; i < len; i += 1) {
    var els = [];
    for (var j = 0; j < numLists; j += 1) {
      els.push(args[j][i]);
    }
    results.push((fn == null) ? els : fn.apply(null, els));
  }
  return results;
};

// zip, zip3, zip4.. all in one!
f.zip = function () {
  return f.zipWith.apply(null, [null].concat(slice.call(arguments, 0)));
};


// comparison

// sort helper
// put in property names (in order), you want to order by
// then pass the resulting function to ary.sort()
f.comparing = function () {
  var pargs = slice.call(arguments, 0);
  return function (x, y) {
    for (var i = 0; i < pargs.length; i += 1) {
      if (x[pargs[i]] !== y[pargs[i]]) {
        return x[pargs[i]] - y[pargs[i]];
      }
      return 0;
    }
  };
};


// functional comparison helpers
f.gt = function (a) {
  return function (b) {
    return b > a;
  };
};

f.lt = function (a) {
  return function (b) {
    return b < a;
  };
};

f.eq = function (a) {
  return function (b) {
    return b === a;
  };
};

// weak eq
f.weq = function (a) {
  return function (b) {
    return b == a;
  };
};

f.gte = function (a) {
  return function (b) {
    return b >= a;
  };
};

f.lte = function (a) {
  return function (b) {
    return b >= a;
  };
};

// can do stuff like
// [1,4,2,5,2,3].filter(gt(3)); -> [4,5]

f.elem = function (ary) {
  return function (x) {
    return ary.indexOf(x) >= 0;
  };
};

f.notElem = function (ary) {
  return function (x) {
    return ary.indexOf(x) < 0;
  };
};

// allows stuff like
// [[1,3,5], [2,3,1]].filter(any(gte(5))) -> [ [ 1, 3, 5 ] ]
// [1,2,3,4,3].filter(elem([1,3]))  -> [1,3,3]
// [1,2,3,4,3].filter(notElem[1,3]) -> [2,4]

// nub, nubBy, intersect, intersectBy?
f.nub = function (ary) {
  var result = [];
  for (var i = 0; i < ary.length; i += 1) {
    if (ary.indexOf(ary[i], i+1) < 0) {
      result.push(ary[i]);
    }
  }
  return result;
};


f.curry = function (fn) {
  var curried = slice.call(arguments, 1);
  return function () {
    var args = curried.concat(slice.call(arguments, 0));
    return fn.apply(this, args);
  };
};

// like curry, but curries the last arguments, and creates a function expecting the first
f.rcurry = function (fn) {
  var curried = slice.call(arguments, 1);
  return function () {
    var args = slice.call(arguments, 0).concat(curried);
    return fn.apply(this, args);
  };
};

// sequence(f1, f2, f3..., fn)(args...) == fn(...(f3(f2(f1(args...)))))
// f.sequence(f.plus(2), f.plus(3), f.times(2))(2) -> 14
f.sequence = function () {
  var fns = slice.call(arguments, 0)
    , argLen = fns.length;
  return function () {
    var result = slice.call(arguments, 0);
    for (var i = 0; i < argLen; i += 1) {
      result = [fns[i].apply(this, result)];
    }
    return result[0];
  };
};


// guard a function by a condition function
// returns a function that will only apply f(x) if cond(x) is true
f.guard = function (fn, cond) {
  return function (x) {
    return (cond(x)) ? fn(x) : null;
  };
};

// var guardedFibonacci = f.guard(fibonacci, lt(100));

// f.either null guard a function, else return errorFn result
// if errorFn is a logger, then curry it with the required message
f.either = function (guardedFn, errorFn) {
  return function (x) {
    var result = guardedFn(x);
    return (result === null) ? errorFn() : result;
  };
};

// var errorMsg;
// var cpuSafeFibonacci = f.either(guardedFibonacci, f.constant(errorMsg));
// or
// var cpuSafeFibonaci = f.either(guardedFibonacci, f.curry(console.log, errorMsg))



f.pow = function (exponent) {
  return function (x) {
    return Math.pow(x, exponent);
  };
};

f.logBase = function (base) {
  return function (x) {
    return Math.log(x) / Math.log(base);
  };
};


// debug function, wrap it in a function reporting its scope and arguments
// particularly useful when combined with f.iterate
f.trace = function (fn, fnName) {
  var log = (console) ? console.log : noop
    , global = (function () { return this; }())
    , name = fn.name || fnName || "fn";

  return function () {
    var result = fn.apply(this, arguments);
    var scope = (this === global) ? "global" : this;
    log('[', name + '.apply(', scope, ',', slice.call(arguments, 0), ') -> ', result, ']');
    return result;
  };
};

module.exports = f;


prelude.js
var slice = Array.prototype.slice;
var p = {};

// common
var id = p.id = function (x) {
  return x;
};

var noop = p.noop = function () {
};

var constant = p.constant = function (val) {
  return function () {
    return val;
  };
};

// curried helper functions for common operations
p.plus = function (x) {
  return function (y) {
    return y + x;
  };
};

p.times = function (x) {
  return function (y) {
    return y * x;
  };
};

p.append = function (xs) {
  return function (ys) {
    return ys.concat(xs);
  };
};

p.prepend = function (xs) {
  return function (ys) {
    return xs.concat(ys);
  };
};

p.compose = function (f) {
  return function (g) {
    return function (x) {
      return f(g(x));
    };
  };
};



// fold - Array::reduce with array curried
var fold = p.fold = function (f, initial) {
  return function (xs) {
    return xs.reduce(f, initial);
  };
};

// binary operators that will be folded over - not exported
var add = function (x, y) {
  return x + y;
};

var multiply = function (x, y) {
  return x * y;
};

var concat = function (xs, ys) {
  return xs.concat(ys);
};

var both = function (x, y) {
  return x && y;
};

var either = function (x, y) {
  return x || y;
};

var chain = function (f, g) {
  return function (x) {
    return f(g(x));
  };
};


// using fold with binary operators to give lifted functions
// any of these (i.e. they act on a list) can be remembered by "take the " fnName
// wheras the variadic counterpart has the name of the associative operator
p.sum = fold(add, 0);
p.product = fold(multiply, 1);
p.flat = fold(concat, []);
p.composition = fold(chain, id);

// we could export these
//p.and = fold(both, true);
//p.or = fold(either, false);

// but prefer faster ES5 methods
// best semantically; "we take the logical" and/or of list
// oh actually, this isn't the same...
p.and = function (f) {
  return function (xs) {
    return xs.every(f);
  };
};

p.or = function (f) {
  return function (xs) {
    return xs.some(f);
  };
};

// lifts and unlifts:
// because some functions are already variadic

// lift a multi param function f into a single array fn, inverse of unlift
var lift = p.lift = function (fn, context) {
  return function (xs) {
    return fn.apply(context, xs);
  };
};

// examples
// follows the "take the" fnName of list semantic
// max/min already have unlifted variadic counterparts so they can simply be lifted
var maximum = p.maximum = lift(Math.max, Math);
var minimum = p.minimum = lift(Math.min, Math);


// (new ClassInst) can't be lifted - has to be one manually
// maybe not want to put this in here, after all Object.create exists
// and classes are a different thing for people to disagree over
p.construct = function (Ctor, args) {
  var F = function () {
    Ctor.apply(this, args);
  };
  F.prototype = Ctor.prototype;
  return new F();
};

// take a function operating on an array and turn it into a multi-parameter function, inverse of lift
// because argument based functions are semantic tand is most sensible with zipWith
var unlift = p.unlift = function (fn, context) {
  return function () {
    var args = slice.call(arguments, 0);
    return fn.apply(context, [args]);
  };
};

// variadic version of lift functions
// named after the operation, i.e. "we " fnName (together) "arg1, arg2, ..."
p.add = unlift(p.sum);
p.multiply = unlift(p.product);
p.concat = unlift(p.flatten);
p.and = unlift(p.all);
p.or = unlift(p.any);
p.compose = unlift(p.composition);


// getters/setters

// simple proprety accessor
// can also be used for array number accessor
// get :: Int/String -> Property
p.get = function (prop) {
  return function (el) {
    return el[prop];
  };
};

// property get map -- equivalent to _.pluck or map(get('prop'))
// zip(pmap('length'), [[1,3,2],[2,1]], [[1],[2]], [[2,3],[2,1]]) -> [[3,1,2], [2,1,2]]
// pmap :: String -> [a] -> [b] -- both curried and as a two param fn
var pmap = p.pmap = function (propName, ary) {
  var fn = function (xs) {
    var result = [];
    for (var i = 0; i < xs.length; i += 1) {
      result[i] = xs[i][propName];
    }
    return result;
  };
  return (ary == null) ? fn : fn(ary);
};

p.set = function (propName) {
  return function (el, value) {
    el[propName] = value;
    return el;
  };
};

// modify a list of objects by setting propName on all objects to valFn(currObj)
// property set map -- equivalent to map(set('prop'))
var mmap = p.mmap = function (propName, valFn) {
  return function (xs) {
    for (var i = 0; i < xs.length; i += 1) {
      xs[i][propName] = valFn(xs[i]);
    }
    return xs;
  };
};

// can use mmap('prop1', constant(5))([{}, {a:2}]) -> [{prop1:5}, {a:2, prop1: 5}]|



// general map/zip helpers

// equivalent to _.range
p.range = function (start, stop, step) {
  if (arguments.length <= 1) {
    stop = start || 0;
    start = 0;
  }
  step = arguments[2] || 1;

  var len = Math.max(Math.ceil((stop - start) / step), 0)
    , idx = 0
    , range = new Array(len);

  while (idx < len) {
    range[idx++] = start;
    start += step;
  }

  return range;
};

p.iterate = function (f, times) {
  return function (x) {
    var result = [x];
    for (var i = 1; i < times; i += 1) {
      result.push(f(result[i - 1]));
    }
    return result;
  };
};


//TODO: throttle, memoize, debounce, once
//TODO: clone, extend

// can act as zipWith, zipWith3, zipWith4...
// zipper function must have the same number of arguments as there are lists
// but beyond that, it's very dynamic
// zipWith(function(x,y,z){return x+y+z;}, [1,3,2], [21,1], [2,3]) -> [24,7]
p.zipWith = function () {
  var fn = arguments[0]
    , args = slice.call(arguments, 1)
    , numLists = args.length
    , results = []
    , len = minimum(pmap('length', args));

  for (var i = 0; i < len; i += 1) {
    var els = [];
    for (var j = 0; j < numLists; j += 1) {
      els.push(args[j][i]);
    }
    results.push((fn == null) ? els : fn.apply(null, els));
  }
  return results;
};

// zip, zip3, zip4.. all in one!
p.zip = function () {
  return p.zipWith.apply(null, [null].concat(slice.call(arguments, 0)));
};


// comparison

// sort helper
// put in property names (in order), you want to order by
// then pass the resulting function to ary.sort()
p.comparing = function () {
  var pargs = slice.call(arguments, 0);
  return function (x, y) {
    for (var i = 0; i < pargs.length; i += 1) {
      if (x[pargs[i]] !== y[pargs[i]]) {
        return x[pargs[i]] - y[pargs[i]];
      }
      return 0;
    }
  };
};


// functional comparison helpers
p.gt = function (a) {
  return function (b) {
    return b > a;
  };
};

p.lt = function (a) {
  return function (b) {
    return b < a;
  };
};

p.eq = function (a) {
  return function (b) {
    return b === a;
  };
};

// weak eq
p.weq = function (a) {
  return function (b) {
    return b == a;
  };
};

p.gte = function (a) {
  return function (b) {
    return b >= a;
  };
};

p.lte = function (a) {
  return function (b) {
    return b >= a;
  };
};

// can do stuff like
// [1,4,2,5,2,3].filter(gt(3)); -> [4,5]

p.elem = function (ary) {
  return function (x) {
    return ary.indexOf(x) >= 0;
  };
};

p.notElem = function (ary) {
  return function (x) {
    return ary.indexOf(x) < 0;
  };
};

// allows stuff like
// [[1,3,5], [2,3,1]].filter(any(gte(5))) -> [ [ 1, 3, 5 ] ]
// [1,2,3,4,3].filter(elem([1,3]))  -> [1,3,3]
// [1,2,3,4,3].filter(notElem[1,3]) -> [2,4]

// nub, nubBy, intersect, intersectBy?
p.nub = function (ary) {
  var result = [];
  for (var i = 0; i < ary.length; i += 1) {
    if (ary.indexOf(ary[i], i+1) < 0) {
      result.push(ary[i]);
    }
  }
  return result;
};


p.curry = function (fn) {
  var curried = slice.call(arguments, 1);
  return function () {
    var args = curried.concat(slice.call(arguments, 0));
    return fn.apply(this, args);
  };
};

// like curry, but curries the last arguments, and creates a function expecting the first
p.rcurry = function (fn) {
  var curried = slice.call(arguments, 1);
  return function () {
    var args = slice.call(arguments, 0).concat(curried);
    return fn.apply(this, args);
  };
};

// sequence(f1, f2, f3..., fn)(args...) == fn(...(f3(f2(f1(args...)))))
// p.sequence(p.plus(2), p.plus(3), p.times(2))(2) -> 14
p.sequence = function () {
  var fns = slice.call(arguments, 0)
    , argLen = fns.length;
  return function () {
    var result = slice.call(arguments, 0);
    for (var i = 0; i < argLen; i += 1) {
      result = [fns[i].apply(this, result)];
    }
    return result[0];
  };
};


// guard a function by a condition function
// returns a function that will only apply f(x) if cond(x) is true
p.guard = function (fn, cond) {
  return function (x) {
    return (cond(x)) ? fn(x) : null;
  };
};

// var guardedFibonacci = p.guard(fibonacci, lt(100));

// p.either null guard a function, else return errorFn result
// if errorFn is a logger, then curry it with the required message
p.either = function (guardedFn, errorFn) {
  return function (x) {
    var result = guardedFn(x);
    return (result === null) ? errorFn() : result;
  };
};

// var errorMsg;
// var cpuSafeFibonacci = p.either(guardedFibonacci, p.constant(errorMsg));
// or
// var cpuSafeFibonaci = p.either(guardedFibonacci, p.curry(console.log, errorMsg))

p.pow = function (exponent) {
  return function (x) {
    return Math.pow(x, exponent);
  };
};

p.logBase = function (base) {
  return function (x) {
    return Math.log(x) / Math.log(base);
  };
};


// debug function, wrap it in a function reporting its scope and arguments
// particularly useful when combined with p.iterate
p.trace = function (fn, fnName) {
  var log = (console) ? console.log : noop
    , global = (function () { return this; }())
    , name = fn.name || fnName || "fn";

  return function () {
    var result = fn.apply(this, arguments);
    var scope = (this === global) ? "global" : this;
    log('[', name + '.apply(', scope, ',', slice.call(arguments, 0), ') -> ', result, ']');
    return result;
  };
};

module.exports = p;
